x/wasm/types/authz.go

```diff
@@ -16,12 +18,92 @@ import (
 const gasDeserializationCostPerByte = uint64(1)
 
 var (
+	_ authztypes.Authorization         = &StoreCodeAuthorization{}
 	_ authztypes.Authorization         = &ContractExecutionAuthorization{}
 	_ authztypes.Authorization         = &ContractMigrationAuthorization{}
 	_ cdctypes.UnpackInterfacesMessage = &ContractExecutionAuthorization{}
 	_ cdctypes.UnpackInterfacesMessage = &ContractMigrationAuthorization{}
 )
 
+// NewStoreCodeAuthorization constructor
+func NewStoreCodeAuthorization(grants ...CodeGrant) *StoreCodeAuthorization {
+	return &StoreCodeAuthorization{
+		Grants: grants,
+	}
+}
+
+// MsgTypeURL implements Authorization.MsgTypeURL.
+func (a StoreCodeAuthorization) MsgTypeURL() string {
+	return sdk.MsgTypeURL(&MsgStoreCode{})
+}
+
+// Accept implements Authorization.Accept.
+func (a *StoreCodeAuthorization) Accept(ctx sdk.Context, msg sdk.Msg) (authztypes.AcceptResponse, error) {
+	var code []byte
+	var permission AccessConfig
+	switch msg := msg.(type) {
+	case *MsgStoreCode:
+		code = msg.WASMByteCode
+		permission = *msg.InstantiatePermission
+	default:
+		return authztypes.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrap("unknown msg type")
+	}
+
+	checksum, err := wasmvm.CreateChecksum(code)
```

Comment:
the binary in `code` is likely zipped. You need to check and unpack as we do in keeper.create before getting the checksum. This would include charging the gas